java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -270,10 +252,23 @@ private predicate constructorStep(Expr tracked, ConstructorCall sink) {
   )
 }
 
+/**
+ * Converts an argument index to a formal parameter index.
+ * This is relevant for varadic methods.
+ */
+private int argToParam(MethodAccess ma, int arg) {
+  exists(ma.getArgument(arg)) and
+  exists(Method m | m = ma.getMethod() |
+    if m.isVarargs() and arg >= m.getNumberOfParameters()
+    then result = m.getNumberOfParameters() - 1
+    else result = arg
+  )
+}
+
 /** Access to a method that passes taint from qualifier to argument. */
 private predicate qualifierToArgumentStep(Expr tracked, Expr sink) {
   exists(MethodAccess ma, int arg |
-    taintPreservingQualifierToArgument(ma.getMethod(), arg) and
+    taintPreservingQualifierToArgument(ma.getMethod(), argToParam(ma, arg)) and
```

Comment:
Based purely on gut feeling, I think the addition of `argToParam` is something that we want to inspect for possible bad join-orders, i.e. evaluate all of this on a big snapshot (jdk11 is likely a good candidate) and check that the tuple counts in the performance log doesn't have any bulges.

Reply:
For the record: We inspected this and everything looked fine.